src/task/executor.rs

```diff
@@ -70,91 +81,166 @@ impl Clone for Kicker {
+                _ => (),
+            }
         }
+        mem::forget(task);
     }
 
-    pub fn resolve(self, success: bool) {
-        // it should always be canceled for now.
-        assert!(success);
-        poll(&Arc::new(self.clone()), true);
+    fn clone_id(&self, id: usize) -> usize {
+        let task = unsafe { Arc::from_raw(id as *mut SpawnTask) };
+        let t = task.clone();
+        mem::forget(task);
+        Arc::into_raw(t) as usize
+    }
+
+    fn drop_id(&self, id: usize) {
+        unsafe { Arc::from_raw(id as *mut SpawnTask) };
     }
 }
 
-impl Notify for SpawnNotify {
-    fn notify(&self, _: usize) {
-        if thread::current().id() == self.worker_id {
-            poll(&Arc::new(self.clone()), false)
-        } else {
-            let mut ctx = self.ctx.lock();
-            if ctx.kicked {
-                return;
-            }
-            ctx.notify(Box::new(CallTag::Spawn(self.clone())));
-            ctx.kicked = true;
-        }
+/// Work that should be differred to be handled.
+///
+/// Sometimes a work can't be done imediately as it might lead
+/// to resourse conflict, deadlock for example. So they will be
+/// pushed into a queue and handled when current work is done.
+pub struct UnfinishedWork(Arc<SpawnTask>);
+
+impl UnfinishedWork {
+    pub fn finish(self, cq: &CompletionQueue) {
+        resolve(cq, self.0, true);
     }
 }
 
 /// Poll the future.
 ///
-/// `woken` indicates that if the cq is kicked by itself.
-fn poll(notify: &Arc<SpawnNotify>, woken: bool) {
-    let mut handle = notify.handle.lock();
-    if woken {
-        notify.ctx.lock().kicked = false;
-    }
-    if handle.is_none() {
-        // it's resolved, no need to poll again.
-        return;
-    }
-    match handle.as_mut().unwrap().poll_future_notify(notify, 0) {
-        Err(_) | Ok(Async::Ready(_)) => {
-            // Future stores notify, and notify contains future,
-            // hence circular reference. Take the future to break it.
-            handle.take();
+/// `woken` indicates that if the cq is waken up by itself.
+fn poll(cq: &CompletionQueue, task: Arc<SpawnTask>, woken: bool) {
+    let mut init_state = if woken { NOTIFIED } else { IDLE };
+    // TODO: maybe we need to break the loop to avoid hunger.
+    loop {
+        match task
+            .state
+            .compare_exchange(init_state, POLLING, Ordering::SeqCst, Ordering::Acquire)
+        {
+            Ok(_) => {}
+            Err(COMPLETED) => return,
+            Err(s) => panic!("unexpected state {}", s),
+        }
+
+        let id = &*task as *const SpawnTask as usize;
+
+        match unsafe { &mut *task.handle.get() }
+            .as_mut()
+            .unwrap()
+            .poll_future_notify(&cq.worker, id)
+        {
+            Err(_) | Ok(Async::Ready(_)) => {
+                task.state.store(COMPLETED, Ordering::SeqCst);
+                unsafe { &mut *task.handle.get() }.take();
+            }
+            _ => {
```

Comment:
`task` will be deallocated in this branch. Is it ok?

Reply:
Yes. `poll_future_notify` will clone it if necessary when it's not ready.